### ISSUE TITLE
Explicit downcast does not work for inherited contracts/interfaces.

### ISSUE THREAD

**Issue by k06a** [ethereum/solidity#12622]
## Description

Explicit downcast does not work for inherited contracts/interfaces.

Imagine you have the following interface hierarchy:
```solidity
interface IFoo {}
interface IBar is IFoo {}
```

We have the following behavior now:
- Upcasts works with **implicit** type casts:

    ```solidity
    IBar bar;
    IFoo foo = bar;
    ```

- Downcast doesn't work even with **explicit** type casts:

    ```solidity
    IFoo foo;
    IBar bar = IBar(foo);
    ```

- We are required to cast over `address`:

    ```solidity
    IFoo foo;
    IBar bar = IBar(address(foo));
    ```

## Environment

- Compiler version: 0.8.11
- Framework/IDE (e.g. Truffle or Remix): Remix, Truffle, Hardhat

**Comment by chriseth** [comment]
I'm not sure we should change the current behaviour. The cast from IFoo to IBar is wrong. It may be "less wrong" than casting to address or casting to a completely unrelated contract, but it is still wrong.

**Comment by k06a** [comment]
@chriseth but this is what happens in a lot of projects and that's why we need **explicit** cast - to show that developer is understanding the risk. Multiple programming languages allow such **explicit** casting, that's why I advocate Solidity should allow too.

**Comment by chriseth** [comment]
Can you provide some pointers to where this conversion across address is used to convert to a derived contract?

**Comment by k06a** [comment]
For example I can't have `IExtendedERC20` instead of `IERC20` in argument:
```solidity
interface IMarket {
    function someAction(IERC20 token, uint256 amount) external;
}

contract MyMarket is IMarket {
    function someAction(IExtendedERC20 token, uint256 amount) external {}
}
```

I am forced to have `IERC20`:
```solidity
function someAction(IERC20 token, uint256 amount) external {
    IExtendedERC20(address(token)).extendedMethod();
}
```

**Comment by ekpyron** [comment]
But even in this example, I'd actually agree with @chriseth - since ``IMarket`` defines the interface function as taking ``IERC20``, every implementation will have to allow for the possibility that ``someAction`` is called with an arbitrary ``IERC20`` token... that's pretty much the promise you give by implementing ``IMarket`` - if my implementation of ``someAction`` will break for anything that's not a ``IExtendedERC20`` token, then it's not really a valid implementation of ``IMarket.someAction``...
Or was this just brevity in the example and you mean something like this?
```solidity
function someAction(IERC20_and_IERC165 token, uint256 amount) external {
  if (token.supportsInterface(type(IExtendedERC20).interfaceId)) {
      IExtendedERC20(address(token)).extendedMethod();
  }
}
```
That's a case in which I guess such a cast would be valid, but since without some additional hoops like this, this is actually dangerous, maybe it's not so bad that this requires a verbose cast in the end...

The danger in a direct explicit cast is that it might suggest that there is actually some kind of type-checking happening there and you might easily be tricked into expecting that the cast reverts if the type doesn't actually match - but that's not the case...

**Comment by k06a** [comment]
My example makes more sense if there is a contract hierarchy instead of interfaces:
```solidity
contract Aminal {}
contract Bird is Animal {}

contract AnimalMover {
    function move(Animal animal, uint256 amount) external {}
}

contract BirdMover is AnimalMover {
    function move(Bird token, uint256 amount) external {}
}
```

I will try to provide a more realistic example soon.

**Comment by k06a** [comment]
Here is a huge discussion on downcasting: https://softwareengineering.stackexchange.com/questions/366592/what-is-a-proper-use-of-downcasting

**Comment by ekpyron** [comment]
I don't doubt that there are situations in which you really want and validly can downcast.
The main advantage of forcing a conversion via ``address`` is that with such a conversion there is no doubt that the cast will *not* involve any kind of runtime type checking or the like, while a direct explicit conversion may suggest otherwise.
But yes, maybe more realistic examples would be helpful.

**Comment by k06a** [comment]
I see such downcasts happens in a lot of different projects (`[^t][. (][A-Z][a-zA-Z0-9]+\(address\([^0]`):

- Balancer https://github.com/balancer-labs/balancer-v2-monorepo

    <img width="626" alt="image" src="https://user-images.githubusercontent.com/702124/152520457-7eb0ff25-5d58-4a2a-a25e-fa625a9954c7.png">

- Compound https://github.com/compound-finance/compound-protocol

    <img width="587" alt="image" src="https://user-images.githubusercontent.com/702124/152520916-96573798-7764-46fe-8953-19ae5474036b.png">

- Aave https://github.com/aave/protocol-v2
    
    <img width="521" alt="image" src="https://user-images.githubusercontent.com/702124/152520608-527a3bc8-7401-483e-885b-898599007b2b.png">

- Euler https://github.com/euler-xyz/euler-contracts

    <img width="473" alt="image" src="https://user-images.githubusercontent.com/702124/152520751-af063c16-eb3f-4ef5-856d-1315fb3f2038.png">

**Comment by k06a** [comment]
There are 3 types of casting:
- Upcast
- Downcast
- Badcast

And there are 3 types of behavior:
- Implicit
- Explicit
- Denied (works over `address`)

Current state:
- Upсasts are implicit
- Downdasts are denied
- Badcasts are denied

Currect difference in behavior for both integral and contract/interface types:

Type | Upcast | Downcast | Badcast
---- | ---- | ---- | ----
integral types | implicit | explicit | denied
contracts/interfaces | implicit | *denied* | denied

I propose making **downcast** to be **explicit**.

**Comment by k06a** [comment]
Right now we can't distinguish syntax of downcast from "badcast". I wish never do "badcast", but Solidity syntax right now do not allow me to check that.

How can I make sure my code has no "badcast"? They look exactly the same way as downcasts.

**I think this is the strongest point in this requirement to allow explicit downcasts.**

**Comment by ekpyron** [comment]
Ok, you do have at the very least some point there.

Conversely, would you say that there is no danger at all that a downcast might suggest that some type checking happens (i.e. that a user or reviewer might be inclined to think that an invalid explicit downcast reverts)? If you go via ``address`` it is much clearer syntactically that no runtime-checking will happen. I guess that's the strongest counterpoint.

**Comment by ekpyron** [comment]
Partly off-topic and mainly to have the issues linked: This is also be something to keep in mind for a future discussion of https://github.com/ethereum/solidity/issues/11284.

**Comment by cameel** [comment]
I think that this is another issue where #11284 is a blocker. When we have that we'll be able to introduce more specific flavors of casts. Downcasts are still dangerous but something like `downcast<IBar>(foo)` would at least check that `IBar` is some specialization of the type that `foo` is, which is at least a bit more than going through `address` gives you.

**Comment by ekpyron** [comment]
https://github.com/ethereum/solidity/issues/11284 is not unlikely to end up being breaking, so I would try to avoid considering it "blocking" for anything (at least too quickly), but rather collect things that can be more smoothly solved by it in 0.9 to consider when we get to it, but still try to solve issues using the current conversions directly, if possible (and there's agreement about it - which we don't yet have in this case here :-)).

**Comment by cameel** [comment]
Well, we know that the current conversion syntax is limiting so I think it's better to do the opposite - first figure out how to fix it. It then makes implementing some features on top of it easy (like downcasting or casting function types). I think that's better than trying to shoehorn it into the current system.

About not having an agreement yet - true. My position is that I see some value in it but I don't think that it's a game changer. Downcast is still a dangerous operation that the compiler cannot fully verify. IMO a downcast syntax would be a slight improvement over a cast to `address` but it also has to be weighed against making the language more complex.

**Comment by frangio** [comment]
I think casts should always be considered unsafe. Integer downcasting is extremely unsafe because it truncates but the compiler doesn't warn about it. Casting "through" address doesn't really provide anything here IMO.

**Comment by k06a** [comment]
I was considering any cast looks like unsafe, but somehow happened, that downcast was denied. Explicit casting itself shows that something strange happening. Now I really afraid if I do “badcasts” anywhere in my code - no way to check my dowcasts.

**Comment by k06a** [comment]
Since upcasts happens implicitly, I assume no-one is using explicit cast operation for upcast.

**Comment by cameel** [comment]
> Integer downcasting is extremely unsafe because it truncates but the compiler doesn't warn about it.

> Since upcasts happens implicitly, I assume no-one is using explicit cast operation for upcast.

These are very good points. I agree, we should just make explicit cast work between arbitrary contracts. They're already unsafe in all but trivial cases because they're only useful with addresses.

**Comment by k06a** [comment]
@cameel by arbitrary contracts you mean downcasting? It will be a bad idea to allow this:
```solidity
interface A {}
interface B {}
A a;
B b = B(a);
```

This "bad" casting should be done via address: `B b = B(address(a));` or better not done at all :)

**Comment by cameel** [comment]
I wanted to say any casting, including to a "bad" contract, because the difference between this and downcasting is not that big to me. Both are inherently unsafe. With downcasting there's one extra check that can catch one specific class of errors but you can still cast to a type not actually supported by the contract.

But, thinking more about it, for example integer conversions mentioned by @frangio don't let you do absolutely anything. For example you have to change sign and width in two separate steps. So in that sense being able to only downcast directly and having to go through `address` to cast to other contract types would not be inconsistent with the rest of the language.

Basically, you convinced me that having `A(b)` do an unsafe conversion would be fine. Whether we should require that unsafe conversion to only be a downcast or if we should allow anything - either one makes sense in some way. I'd like to hear what others think on this point.

**Comment by k06a** [comment]
@cameel I wish compiler to prevent me from casting `A(b)` when contract/interfaces are siblings or even not in the same hierarchy. If I want to make sure this is not happening I will just avoid using `A(address(b))` ever - this will let me believe I have only implicit upcasts and explicit downcasts.

**Comment by k06a** [comment]
So what about downcasting a bit more safe than “bad” casting out of any straight hierarchy or “sibling” casting?
